File: 1937.扣分后的最大得分.js
/*
 * @lc app=leetcode.cn id=1937 lang=javascript
 *
 * [1937] 扣分后的最大得分
 */

// @lc code=start
/**
 * @param {number[][]} points
 * @return {number}
 */
var maxPoints = function(points) {
    let m = points.length;
    let n = points[0].length;
    let dp = new Array(n).fill(0);
    
    // 一开始，同位置时，最大值是上一行的值。
    // 偏移时，马上确定当前列的最大值。且其中一个比较数是可以随位移变小的，像梦琪一样。
    for (let r = 0; r < m; r++) {
        let cur = new Array(n);
        let lmax = 0;
        // 从左向右遍历，以最大值位置向右递减，与本身的值比较替换为较大值
        for (let c = 0; c < n; c++) {
            lmax = Math.max(lmax - 1, dp[c]);
            cur[c] = lmax;
        }
        
        let rmax = 0;
        // 从右向左遍历，以最大值位置向左递减，与本身的值比较替换为较大值
        for (let c = n - 1; c >= 0; c--) {
            rmax = Math.max(rmax - 1, dp[c]);
            cur[c] = Math.max(cur[c], rmax);
        }
        // 然后与当前行的值相加（同列相加）
        for (let c = 0; c < n; c++) {
            dp[c] = cur[c] + points[r][c];
        }
    }
    
    return Math.max(...dp);
}
// @lc code=end

File: 1937.扣分后的最大得分.rs
/*
 * @lc app=leetcode.cn id=1937 lang=rust
 *
 * [1937] 扣分后的最大得分
 */

// @lc code=start
use std::cmp::max;
impl Solution {
    pub fn max_points(points: Vec<Vec<i32>>) -> i64 {
        let m = points.len();
        let n = points[0].len();
        let mut dp: Vec<i64> = vec![0 as i64; n];
        for r in 0..m {
            let mut cur = vec![0; n];
            let mut lmax = 0;
            for c in 0..n {
                lmax = max(lmax - 1, dp[c]);
                cur[c] = lmax;
            }

            let mut rmax = 0;
            for c in (0..n).rev() {
                rmax = max(rmax - 1, dp[c]);
                cur[c] = max(rmax, cur[c]);
            }

            for c in 0..n {
                dp[c] = cur[c] + points[r][c] as i64;
            }
        }

        *dp.iter().max().unwrap()
    }
}
// @lc code=end
